Third-party JS wrap up (#1276)
GoogleChrome/web.dev: src/site/content/en/fast/codelab-optimize-third-party-javascript/index.md

* A data-visualization library for rendering a line graph

* A social media sharing widget

You'll start by measuring the performance of the app and then apply each technique to improve different aspects of app performance.

![Screenshot of the page with third-party resources highlighted.](app.png)

## Measure performance

{% Instruction 'preview' %}

Run a [Lighthouse](https://developers.google.com/web/tools/lighthouse/) [performance audit](/lighthouse-performance) on the page to establish baseline performance:

{% Instruction 'devtools-audits', 'ol' %}
1. Click **Mobile**.
1. Select the **Performance** checkbox. (You can clear the rest of the checkboxes in the Audits section.)
1. Click **Simulated Fast 3G, 4x CPU Slowdown**.
1. Select the **Clear Storage** checkbox.
1. Click **Run audits**.

When you run an audit on your machine, [the exact results may vary](https://developers.google.com/web/tools/lighthouse/variability), but you should notice that the [First Contentful Paint](/first-contentful-paint/) (FCP) time is pretty high, and Lighthouse suggests two opportunities to investigate: **Eliminate render-blocking resources** and **Preconnect to required origins**.

> ![Screenshot of Lighthouse audit showing 2.4 second FCP and two opportunities: Eliminate render-blocking resources and Preconnect to required origins.](lighthouse-audit.png)

Author: I am aware, the variability is wild. The perf score depends on the metrics (FCP, TTI...) and not particular audits that come up in diagnostics/opportunities. Still, following the instructions from this post results in improved metrics and that's what counts?

> ![Screenshot of Lighthouse audit showing 2.4 second FCP and two opportunities: Eliminate render-blocking resources and Preconnect to required origins.](lighthouse-audit.png)

Author: Elaborating on the variability would kinda derail the tutorial, but maybe we can make the disclaimer stand out more?

> ![Screenshot of Lighthouse audit showing 2.4 second FCP and two opportunities: Eliminate render-blocking resources and Preconnect to required origins.](lighthouse-audit.png)

Reviewer: I think that idea sounds good--and maybe include the thought above in the codelab itself: that the improvement is what counts.

## Defer third-party JavaScript

> The **Eliminate render-blocking resources** audit identified that you can save approximately 1.6 seconds by deferring a script coming from d3js.org.

Author: I included it because it was the one data point that had the smallest deviation :D

> The **Eliminate render-blocking resources** audit identified that you can save approximately 1.6 seconds by deferring a script coming from d3js.org.

Reviewer: OK :)

```

### Step 2: Ensure the correct order of operations

Now that D3 is deferred, `script.js` will run before D3 is ready, resulting in an error.

Scripts with the `defer` attribute execute in the order in which they were specified. To ensure `script.js` gets executed after D3 is ready, add `defer` to it and move it up to the `<head>` of the document, right after the D3 `<script>` element. Now it no longer blocks the parser, and the download starts sooner.

```html
<script src="https://d3js.org/d3.v3.min.js" defer></script>
<script src="./script.js" defer></script>
```

## Lazy-load third-party resources

All resources that are below the fold are good candidates for [lazy-loading](/efficiently-load-third-party-javascript/#lazy-load-third-party-resources). 

The sample app has a YouTube video embedded in an [iframe](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe). To check out how many requests the page makes and which come from the embedded YouTube iframe:

{% Instruction 'devtools-network', 'ol' %}
{% Instruction 'disable-cache', 'ol' %}
1. Select *Fast 3G* in the **Throttling** dropdown menu.
{% Instruction 'reload-page', 'ol' %}

> ![Screenshot of DevTools Network panel.](network.png)

Author: I'll get to it a bit later 👍

The **Network** panel reveals that the page made a total of 28 requests and loaded almost 1 MB of compressed resources.

> To identify the requests that the YouTube `iframe` made, look for the video ID `6lfaiXM6waw` in the **Initiator** column. To group together all the requests by domain:

Author: Domain sorting is just a short-cut in this case because it's easy to connect the dots: there's a bunch of Google domains -> ofc that YouTube called them. 

Initiator column is the only way to follow the trail of propagated calls: 

* `initiator:` 6lfaiXM6waw (youtube.com) loads `resource:` base.js (also youtube.com)
* `initiator:` base.js (youtube.com) loads `resource:` ads.js (from doubleclick.google.com)
* etc.

If it was any other third-party embed the grouping would be not be so clear when sorted by domain.

I understand now that this not at all obvious from the codelab, but I really don't want to drag out that step.

### Step 1: Prevent video from loading initially

To lazy-load the video iframe, you must first prevent it from loading in the usual way. Do that by replacing the `src` attribute with the `data-src` attribute to specify the video URL:

```html
<iframe width="560" height="315" data-src="https://www.youtube.com/embed/lS9D6w1GzGY" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```

`data-src` is a [data attribute](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes), which allows you to store extra information on standard HTML elements. A data attribute can be named anything, as long as it starts with "data-". 

An iframe without a `src` simply won't load. 

### Step 2: Use Intersection Observer to lazy-load the video

To load the video when a user scrolls to it you need to know when that happens. That’s where the Intersection Observer API steps in. The Intersection Observer API lets you register a callback function that's executed whenever an element you want to track enters or exits the viewport.

To get started, create a new file and name it `lazy-load.js`:

> {% Instruction 'create' %}

Author: I think it's okay because the UI won't let you _not name it_ and it's fairly obvious where the file name goes. 

We could also say "Click New File and give it a name." but do we really want to err on the side of overexplaining in this straightforward case?

> {% Instruction 'create' %}

Reviewer: Yeah, sorry to be a pain, but I do think enumerated steps need to be fully accurate. I'm OK with the solution you propose so we can preserve this as a shortcode :)

> {% Instruction 'create' %}

Author: cc @robdodson just an FYI that I'll need to update Instruction.js

```js
// create a new Intersection Observer
let observer = new IntersectionObserver(callback);
```

Now give `observer` a target element to watch (the video iframe in this case) by passing it as an argument in the `observe` method:

```js
// the element that you want to watch
const element = document.querySelector('iframe');

// register the element with the observe method
observer.observe(element);
```

`callback` receives a list of [`IntersectionObserverEntry`](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserverEntry) objects and the [`IntersectionObserver`](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver) object itself. Each entry contains a [`target`](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserverEntry/target) element and properties that describe its dimensions, position, the time it entered the viewport, and more. One of the properties of `IntersectionObserverEntry` is [`isIntersecting`](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserverEntry/isIntersecting)—a boolean value that equals `true` when the element enters the viewport.

> In this example, the `target` is the `iframe`. The [`isIntersecting`](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserverEntry/isIntersecting) equals `true` when `target` enters the viewport. To see this in action, replace `callback` with the following function: 

Reviewer: ```suggestion
In this example, the `target` is the `iframe`. `isIntersecting` equals `true` when `target` enters the viewport. To see this in action, replace `callback` with the following function: 
```
- No need for definite article without a noun.
- No need to link it more than once.

```js/1-7/0
let observer = new IntersectionObserver(callback);
let observer = new IntersectionObserver(function(entries, observer) { 
    entries.forEach(entry => {
      console.log(entry.target);
      console.log(entry.isIntersecting);
    });
  });
```
{% Instruction 'preview', 'ol' %}
{% Instruction 'devtools-console', 'ol' %}

> Try scrolling up and down. You should see the value of `isIntersecting` change and the target element logged to the console.

Reviewer: This isn't working for me.

Author: What, lazy loading? And in the final version that I've linked below?

Author: Tried again in a remixed Glitch, worked for me. If you can dig up a link to your remix, I can take a look to see if there's an issue.

Reviewer: Just tried it again, but still not working: https://glitch.com/~hospitable-trigonometry

To load the video when the user scrolls to its position, use `isIntersecting` as a condition to run a `loadElement` function, which gets the value from the `iframe` element’s `data-src` and sets it as the `iframe` element's `src` attribute. That replacement triggers the loading of the video. Then, once the video is loaded, call the [`unobserve`](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver/unobserve) method on the `observer` to stop watching the target element:

> ```js/4-11,13-16/2-3

Reviewer: ```suggestion
```js/6-13,15-18/2-5
```
To compensate for line shift caused by edit below.

let observer = new IntersectionObserver(function (entries, observer) {
  entries.forEach(entry => {
    console.log(entry.target);

>     console.log(entry.isIntersecting);

Reviewer: ```suggestion
    console.log(entry.isIntersecting);
  });
});
```
Needed since the later code highlighting indicates these lines are added.

    if (entry.isIntersecting) {
      // do this when the element enters the viewport
      loadElement(entry.target);
      // stop watching
      observer.unobserve(entry.target);
    }
  });
});

function loadElement(element) {
  const src = element.getAttribute('data-src');
  element.src = src;
}
```

Finally, add `lazy-load.js` to your document head:

 ```html
 <script src="/lazy-load.js" defer></script>
 ```

### Step 3: Reevaluate performance

To see how the size and number of resources changed, open the DevTools **Network** panel and reload the page again. The **Network** panel reveals that the page made 14 requests and only 260 KB. That's a meaningful improvement!

Now scroll down the page and keep an eye on the **Network** panel. When you get to the video, you should see the page trigger additional requests.

<video autoplay="" loop="" muted="" playsinline="">
    <source src="./lazy-load-3g.mp4" type="video/mp4">
  </video>

{% Aside %}

> In this example, the element is loaded when it enters the viewport and you can see it happening in the video above. You can avoid that delay and create smoother user-experience by starting to load elements a little before they enter the viewport. Use [`rootMargin`](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API#Creating_an_intersection_observer) property to define margins around the target element which effectively grow (or shrink) the area that triggers the `isIntersecting` change.

Reviewer: ```suggestion
In this example, the element is loaded when it enters the viewport, which you can see happening in the video above. You can avoid that delay and create a smoother user experience by loading elements a little before they enter the viewport. To do that, use the [`rootMargin`](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API#Creating_an_intersection_observer) property to define margins around the target element, which effectively grows (or shrinks) the area that triggers the `isIntersecting` change.
```

{% endAside %}

## Preconnect to required origins

You’ve deferred non-critical Javascript and lazy-loaded the YouTube requests, so now it’s time to optimize the remaining third-party content.

Adding the `rel=preconnect` attribute to a link tells the browser to establish a connection to a domain before the request for that resource is made. This attribute is best used on origins that provide resources you are certain the page needs. 

The Lighthouse audit you ran in the first step suggested in **Preconnect to required origins** that you can save around 400 ms by [establishing early connections](/preconnect-and-dns-prefetch/) to staticxx.facebook.com and youtube.com:

![Preconnect to required origins audit with the staticxx.facebook.com domain highlighted.](preconnect.png)

Since the YouTube video is now lazy-loaded, that leaves only staticxx.facebook.com, the source of the social media sharing widget. Establishing an early connection to this domain is as simple as adding a `<link>` tag to the document's `<head>`:

```html  
  <link rel="preconnect" href="https://staticxx.facebook.com">
```

## Reevaluate performance 

> Here's the state of the page [after optimization](https://glitch.com/~3rd-party-optimizations). Follow the steps from the [Measure performance](#measure-performance) section of the codelab to run another Lighthouse audit. 

Reviewer: You know, I don't think you need the link to the final version of the app. The reader has presumably been following along--which the rest of the codelab assumes--so I think we should continue assuming that.

Author: I know it's not necessary and we're usually not including it, but I kinda like to have it as a reference at the end in case people made missteps. Your call, just an FYI that it wouldn't be a precedent since I've included it in one of my previous codelabs too :)

Author: This is not a blocker, I'm cool with merging either way.

Reviewer: Fair enough. I rephrased the paragraph a bit because it's unclear which Glitch you're supposed to run the audit on, which is I think what's really bothering me.

![Lighthouse audit showing 1 second FCP and the performance score of 99.](success.png)

{% Aside 'success' %}

> You should see that FCP and the performance score have improved! 🎉 The page now loads faster, consumes less data, and all without negatively affecting the user-experience. 🤝

Reviewer: ```suggestion
You should see that FCP and the performance score have improved! 🎉 The page now loads faster and consumes less data⁠—and all without negatively affecting the user-experience. 🤝
```
The "and all" construction is usually treated as an aside rather than the last item in a list.

{% Instruction 'devtools-audits', 'ol' %}
1. Click **Mobile**.
1. Select the **Performance** checkbox. (You can clear the rest of the checkboxes in the Audits section.)
1. Click **Simulated Fast 3G, 4x CPU Slowdown**.
1. Select the **Clear Storage** checkbox.
1. Click **Run audits**.

When you run an audit on your machine, [the exact results may vary](https://developers.google.com/web/tools/lighthouse/variability), but you should notice that the [First Contentful Paint](/first-contentful-paint/) (FCP) time is pretty high, and that Lighthouse suggests two opportunities to investigate: **Eliminate render-blocking resources** and **Preconnect to required origins**. (Even if the metrics are all in the green, these optimizations will still yield improvements.)

<figure class="w-figure">
  <img class="w-screenshot" src="./lighthouse-audit.png" alt="Screenshot of Lighthouse audit showing 2.4 second FCP and two opportunities: Eliminate render-blocking resources and Preconnect to required origins.">
</figure>

## Defer third-party JavaScript

> The **Eliminate render-blocking resources** audit identified that you can save some time by deferring a script coming from d3js.org.

Reviewer: ```suggestion
The **Eliminate render-blocking resources** audit identified that you can save some time by deferring a script coming from d3js.org:
```

> ![Screenshot of Eliminate render-blocking resources audit with the d3.v3.min.js script highlighted.](render-blocking.png)

Reviewer: ```suggestion
<figure class="w-figure">
  <img class="w-screenshot" src="./render-blocking.png" alt="Screenshot of Eliminate render-blocking resources audit with the d3.v3.min.js script highlighted.">
</figure>
```

{% endAside %}

## Preconnect to required origins

You've deferred non-critical Javascript and lazy-loaded the YouTube requests, so now it's time to optimize the remaining third-party content.

Adding the `rel=preconnect` attribute to a link tells the browser to establish a connection to a domain before the request for that resource is made. This attribute is best used on origins that provide resources you are certain the page needs. 

The Lighthouse audit you ran in the first step suggested in **Preconnect to required origins** that you can save around 400 ms by [establishing early connections](/preconnect-and-dns-prefetch/) to staticxx.facebook.com and youtube.com:

![Preconnect to required origins audit with the staticxx.facebook.com domain highlighted.](preconnect.png)

Since the YouTube video is now lazy-loaded, that leaves only staticxx.facebook.com, the source of the social media sharing widget. Establishing an early connection to this domain is as simple as adding a `<link>` tag to the document's `<head>`:

```html  
  <link rel="preconnect" href="https://staticxx.facebook.com">
```

## Reevaluate performance 

> Here's the state of the page [after optimization](https://glitch.com/~3rd-party-optimizations). Follow the steps from the [Measure performance](#measure-performance) section of the codelab to run another Lighthouse audit. 

Reviewer: ```suggestion
Follow the steps from the [Measure performance](#measure-performance) section of the codelab to run another Lighthouse audit on the sample app. (Here's a [version with all the optimizations](https://glitch.com/~3rd-party-optimizations) if you want to check your work.)
```
